Guard against allowing ansible to ansible-base upgrades

sivel opened the pull request:
##### SUMMARY
Guard against allowing ansible to ansible-base upgrades

There doesn't seem to be a great way for erroring in `setup.py` without generating some form of exception. `RuntimeError` is used elsewhere, so I've used it here too.

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
setup.py

##### ADDITIONAL INFORMATION
The output is roughly the same from pip 9 through pip 20:

```
    ERROR: Command errored out with exit status 1:
     command: /Users/matt/venvs/tmp-3c6561a2d73f68d4/bin/python -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/Users/matt/projects/ansibledev/ansible/setup.py'"'"'; __file__='"'"'/Users/matt/projects/ansibledev/ansible/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' egg_info --egg-base /private/tmp/pip-pip-egg-info-mthvusdc
         cwd: /Users/matt/projects/ansibledev/ansible/
    Complete output (27 lines):
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/Users/matt/projects/ansibledev/ansible/setup.py", line 108, in <module>
        _validate_install_ansible_base()
      File "/Users/matt/projects/ansibledev/ansible/setup.py", line 75, in _validate_install_ansible_base
        raise RuntimeError(
    RuntimeError:

        ****************************************************************************

        Cannot install ansible-base with a pre-existing ansible==2.9.10
        installation.

        Installing ansible-base with ansible-2.9 or older currently installed with
        pip is known to cause problems. Please uninstall ansible and install the new
        version:

            pip uninstall ansible
            pip install ansible-base

        If you want to skip the conflict checks and manually resolve any issues
        afterwards, set the ANSIBLE_SKIP_CONFLICT_CHECK environment variable:

            ANSIBLE_SKIP_CONFLICT_CHECK=1 pip install ansible-base

        ****************************************************************************

    ----------------------------------------
ERROR: Command errored out with exit status 1: python setup.py egg_info Check the logs for full command output.
```

sivel committed:
Guard against allowing ansible to ansible-base upgrades

sivel committed:
newline

sivel committed:
use alias

nitzmahone commented on setup.py:
```diff
@@ -23,6 +23,39 @@
+        if version_tuple < (2, 10):
+            stars = '*' * 76
+            raise RuntimeError(
+                '\n\n%s\n\nCannot upgrade ansible==%s to ansible-base. You must first uninstall\n'
```
Maybe add "ansible versions below 2.10" or something so it's clear this is a one-time problem, not that they can never upgrade `ansible` again...

sivel replied:
I added an extra sentence explicitly detailing this.  Still open to any other things.

sivel committed:
Add an explicit line detailing this is a 1 time thing

sivel committed:
period

sivel committed:
Read __version__ and __author__ rather than import, update working, and add ability to skip conflict checks

sivel commented:
@nitzmahone I added the ENV var for skipping checks, updated the error, and changed how we were getting `__author__` and `__version__`.

I couldn't get `--global-option` to work as expected, so I abandoned it.

Let me know your thoughts.

webknjaz commented on setup.py:
```diff
@@ -23,8 +25,87 @@
+    # Open in Latin-1 so that we avoid encoding errors.
+    # Use codecs.open for Python 2 compatibility
+    try:
+        f = codecs.open(os.path.join(*file_paths), 'r', 'latin1')
+        info_file = f.read()
+        f.close()
```
Any reason not to use a CM?
```suggestion
        with codecs.open(os.path.join(*file_paths), 'r', 'latin1') as f:
            info_file = f.read()
```

sivel replied:
Sorry, that was a copy/paste from pypa, and then slightly modified.

webknjaz commented on setup.py:
```diff
@@ -23,8 +25,87 @@
+                # 'from ansible<2.10 to ansible-base.\n\n%s\n' % (stars, __version__, stars)
+            )
+    finally:
+        sys.path[:] = sys_path
+        for key in sys_modules_keys.symmetric_difference(sys.modules):
+            sys.modules.pop(key, None)
+        sys.modules.update(sys_modules)
```
What is this supposed to achieve? Modules are cached on import, they'll be picked up from that cache regardless of the pointers being listed in `sys.modules`.

sivel replied:
This is just setting things back to the way they were, the `.update()` is likely pointless as it stands, but it was to safe guard some future state where the act of importing a module, has the side effect of deleting some other module from `sys.modules`.

webknjaz commented on setup.py:
```diff
@@ -1,6 +1,8 @@
 from __future__ import (absolute_import, division, print_function)
 __metaclass__ = type
 
+import ast
```
Unused import
```suggestion
```

sivel replied:
Oops, that was left from doing an `ast.literal_eval` on the `os.getenv` but I removed it to avoid any tracebacks that could come from invalid data.

abadger replied:
Still present ;-)

webknjaz commented:
> I couldn't get `--global-option` to work as expected, so I abandoned it.

You probably want `--install-option` instead. The global one is passed to all requested packages in `pip install ...` while `--install-option` is specific to one dist only (hence the position in CLI args matter).
AFAIR you'd have to inherit the command class to access these.

webknjaz commented:
Oh, and using env var is a good alternative. In fact, it's better because you `--install-option` is not whitelisted for use in `requirements.txt` files, only in actual CLI. This is why we had to use env vars in projects like aiohttp/multidict/yarl.

sivel commented on setup.py:
```diff
@@ -23,8 +24,86 @@
+    try:
+        with codecs.open(os.path.join(*file_paths), 'r', 'latin1') as f:
+            info_file = f.read()
+    except:
```
```suggestion
    except Exception:
```

sivel commented on setup.py:
```diff
@@ -23,8 +24,86 @@
+    """Validate that we can install ansible-base. Currently this only
+    cares about upgrading to ansible-base from ansible<2.10
+    """
+    if os.getenv('ANSIBLE_SKIP_CONFLICT_CHECK', '') not in ('', 0):
```
```suggestion
    if os.getenv('ANSIBLE_SKIP_CONFLICT_CHECK', '') not in ('', '0'):
```

sivel committed:
Remove commented code

sivel committed:
Re introduce removed changes from rebase

abadger commented on setup.py:
```diff
@@ -30,8 +32,83 @@
+    # Open in Latin-1 so that we avoid encoding errors.
+    # Use codecs.open for Python 2 compatibility
+    try:
+        with codecs.open(os.path.join(*file_paths), 'r', 'latin1') as f:
```
Latin1 seems wrong...
* In python 3, the default encoding of python files is utf8. 
* When we explicitly designate for python 2, we use utf-8

If you don't want to trust that this is utf-8 (but i think we should), then we should use bytes to read and convert that to text at some point before returning.

webknjaz replied:
That is true but also the target file only contains ASCII so this will not fail as long as we don't "add emojis" there at which point it'll fail in the CI. So there's a really low risk of shipping broken dists...

sivel replied:
This came directly from pypa, so I'm just using their recommendation.  Their version has changed in recent years where py2 doesn't matter.  But I didn't write any of this.

https://github.com/pypa/sampleproject/commit/ec93951fcf42c85e1b77d8b8968d6001345cf820#diff-2eeaed663bd0d25b7e608891384b7298

webknjaz replied:
Yeah, that's from 2013. Nowadays it probably matters less. If only we had less weird things in setup.py, we could just migrate to setup.cfg mostly...

sivel replied:
Here is the current version of that recommendation: https://packaging.python.org/guides/single-sourcing-package-version/

abadger replied:
<nod> okay,.  Since we control the release.py  file, we probably shouldn't add this particular encoding code, then.  It overcomplicates things and confused people who don't have encoding experience won't understand why latin1 is used here and will assume the file actually contains characters encoded in Latin 1). I suggest either ignoring the encoding question if we are going to say the release.py file is ascii-only or reading as bytes and then decoding from utf8 if we want to future proof it against things like a non-ascii code name. (Importing and using codecs is another thing that will be nonobvious to people who are unfamiliar with encoding issues)

abadger replied:
(the pypa recommendation is written to work with any project under any conditions which makes it overly complex for most projects actually implementing it.)

abadger commented on setup.py:
```diff
@@ -30,8 +32,83 @@
+    """Validate that we can install ansible-base. Currently this only
+    cares about upgrading to ansible-base from ansible<2.10
+    """
+    if os.getenv('ANSIBLE_SKIP_CONFLICT_CHECK', '') not in ('', '0'):
```
Hmmn... I think we are going to want the env var here and in the ansible package to match.  Otherwise users will have to set both if they want to override.  If you prefer your name, just open a pr against ansible-community/antsibull to change that one

sivel replied:
That was the plan. We discussed the env var, and preferred this one. Once this merged I planned to do as you mention.

abadger replied:
<nod> please open that pr sooner, rather than later (aka, as soon as you are reasonably certain this is the env var name you want.). It is user facing so it would be good to expose the minimal number of people possible to the old name

sivel committed:
Just use open

sivel committed:
Nuke unused import
